connection_plugins/machinectl.py

```diff
@@ -574,13 +574,27 @@ def put_file(self, in_path, out_path):
         if not os.path.exists(to_bytes(in_path, errors='strict')):
             raise AnsibleFileNotFound('file or module does not exist: {0}'.format(in_path))
 
+        # Okay, this is definitely not a great idea to do, that's pretty ugly,
+        # but we have no choice... Let me explain
+        # You cannot "copy-to --force" with machinectl. There is a request to
+        # do that, but unaddressed as of today:
+        #   https://github.com/systemd/systemd/issues/9441
+        # So you cannot overwrite an existing file. This is very annoying when
+        # pushing DIRECTORIES... as the same ansible file (with the same name
+        # on the remote target) must be overwritten for each file in the
+        # directory.
+        # Without removing the file first, we get an error: "file exists".
+        returncode, stdout, stderr = self._run_command('shell', args=['/bin/sh', '-c', f'rm -rf "{out_path}"'], machine=self.machine)
```

Comment:
Although it's probably safe to assume that a `systemd-machined` is Linux (or at least POSIX-compliant), I would prefer something like the following:

```
remove_cmd = self._shell.remove(out_path, recurse=True)
returncode, stdout, stderr = self._run_command('shell', args=remove_cmd, machine=self.machine)
```
This respects customization the user has done with respect to shell selection (e.g. setting [`ansible_shell_executable`](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html#ansible-shell-executable)).

Additionally, `_shell.remove` handles quoting the `path` argument, if necessary.

Reply:
Hmm, darn.  Couple of subtleties/hurdles:

1. `self._shell.remove` returns a command string; needs to be run through `shlex.split` before being run through `self._run_command`.
2. `self._shell.remove` returns an unqualified path for `rm`, and `machinectl shell` requires an absolute path.

Thinking a combination of these approaches is maybe the way to go:

```
remove_cmd = self._shell.remove(out_path, recurse=True)
remove_args = [self._play_context.executable, '-c', remove_cmd]
returncode, stdout, stderr = self._run_command('shell', args=remove_args, machine=self.machine)
```

Though this assumes that `self._play_context.executable` recognizes the `-c` option.  

Thoughts?

Reply:
To the best of my knowledge, there is no support for systemd-* on non-linux OSes (that are still maintained). 
But I agree that trying to embrace Ansible's abstraction as much as possible is desirable. I'll update my code.

There is indeed a problem with the path for `rm`, I haven't found better than your solution: ask for the shell (managed by Ansible) to execute the command.  I can see in [Ansible's code](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/shell/sh.py) that `sh` shall be POSIX-compliant; fortunately it seems that `-c` is [standardized and required](https://pubs.opengroup.org/onlinepubs/9699919799/).
I guess this is a reasonable assumption.